internal/dnfjson/dnfjson.go

```diff
@@ -102,13 +104,29 @@ type Solver struct {
 	// system and required for subscription support.
 	releaseVer string
 
+	// Full distribution string, eg. fedora-38, used to create separate dnf cache directories
+	// for each distribution.
+	distro string
+
 	subscriptions *rhsm.Subscriptions
 }
 
 // Create a new Solver with the given configuration. Initialising a Solver also loads system subscription information.
-func NewSolver(modulePlatformID string, releaseVer string, arch string, cacheDir string) *Solver {
+func NewSolver(modulePlatformID, releaseVer, arch, distro, cacheDir string) *Solver {
```

Comment:
My original thought on this issue was to have multiple solvers attached to the API, each initialised with a different cache root (one for each distro) under the main cache directory.  Changing the `Solver` to manage them itself is certainly cleaner, but I'm not sure how the cache management is affected now.  The `rpmCache` assumes that the top-level elements in it are rpm repository files and directories.  Now, they're distro subdirectories with repo files and directories under those.  In other words, the `rpmCache.repoElements` are now distro elements instead.

If you look at `updateCacheDirInfo`, you'll see this assumption is used to group elements per repo based on filename: https://github.com/osbuild/osbuild-composer/blob/7741fce11bd114ee950e6d8dd47e1950c1ee3b2f/internal/dnfjson/cache.go#L102-L112

The 64 characters are the hash we compute for the repository and depend on it to identify unique repos and reuse the same directories.  I think with this change to distros, the cleanup code will basically consider the distro subdirectories as "unknown" and ignore them.

We should either have separate `rpmCache` objects for each distro or, if we want to keep the cache management centralised, change the cleanup and file tracking code to work with the distro assumption.  With the latter solution, we could also solve our leftover cache issue by detecting and deleting top-level repo entries on cache creation.

Reply:
That new function is called from a loop that iterates over the top level directories, the name being hashed should still be the same as it was before. But I'm not sure if my attempt works, I think I need to add some tests, or some prints at least :)

Reply:
I've added a function (`CleanupOldCacheDirs`) that is called right before the initial metadata download and cache setup. It removes anything under ./rpmmd/ that it doesn't know about. Including old top-level cache files and directories, and old distros that have been removed from `supportedDistros`.